KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,652 @@
+
+        emit NewbieRemoved(_newbieAddress, newbies.length);
+    }
+
+    /**
+     * @dev retiredAddress can be a EOA or a contract address. To approve:
+     *      If the retiredAddress is a EOA, the msg.sender should be the EOA address
+     *      If the retiredAddress is a Contract, the msg.sender should be one of the contract `admin`.
+     *      It uses the getState() function in the retiredAddress contract to get the admin details.
+     * @param _retiredAddress is the address of the retired
+     */
+    function approve(
+        address _retiredAddress
+    ) public onlyAtStatus(Status.Registered) {
+        require(
+            retiredExists(_retiredAddress),
+            "retired needs to be registered before approval"
+        );
+
+        //Check whether the retired address is EOA or contract address
+        bool isContract = isContractAddr(_retiredAddress);
+        if (!isContract) {
+            //check whether the msg.sender is the retired if its a EOA
+            require(
+                msg.sender == _retiredAddress,
+                "retiredAddress is not the msg.sender"
+            );
+            _updateApprover(_retiredAddress, msg.sender);
+        } else {
+            //check if the msg.sender is one of the admin of the retiredAddress contract
+            require(
+                _validateAdmin(_retiredAddress, msg.sender),
+                "msg.sender is not the admin"
+            );
+            _updateApprover(_retiredAddress, msg.sender);
+        }
+    }
+
+    /**
+     * @dev validate if the msg.sender is admin if the retiredAddress is a contract
+     * @param _retiredAddress is the address of the contract
+     * @param _approver is the msg.sender
+     * @return isAdmin is true if the msg.sender is one of the admin
+     */
+    function _validateAdmin(
+        address _retiredAddress,
+        address _approver
+    ) private returns (bool isAdmin) {
+        (address[] memory adminList, ) = _getState(_retiredAddress);
+        require(adminList.length != 0, "admin list cannot be empty");
+        for (uint8 i = 0; i < adminList.length; i++) {
+            if (_approver == adminList[i]) {
+                isAdmin = true;
+            }
+        }
+    }
+
+    /**
+     * @dev gets the adminList and quorom by calling `getState()` method in retiredAddress contract
+     * @param _retiredAddress is the address of the contract
+     * @return adminList list of the retiredAddress contract admins
+     * @return req min required number of approvals
+     */
+    function _getState(
+        address _retiredAddress
+    ) private returns (address[] memory adminList, uint256 req) {
+        //call getState() function in retiredAddress contract to get the adminList
+        bytes memory payload = abi.encodeWithSignature("getState()");
+        (bool success, bytes memory result) = _retiredAddress.staticcall(
+            payload
+        );
+        emit GetState(success, result);
```

Comment:
I think this is not useful, because if success==false, this event will not emitted because the transaction will fail.

Reply:
But if success=true, It would good to log the result which is the adminList. Just in case of error, we can decode what was the admin List.

Reply:
@toniya-klaytn Yes, for that reason, it could be better to make it more precise, like RetiredAddrAdminStatus. I think this already changed as we discussed. I will make this conversation as resolved.